packages/SwingSet/src/kernel/kernel.js

```diff
@@ -999,135 +1166,98 @@ export default function buildKernel(
       message,
     });
     /** @type { PolicyInput } */
-    let policyInput = ['none'];
+    let policyInput = ['crank', {}];
     if (message.type === 'create-vat') {
       policyInput = ['create-vat', {}];
     }
 
-    // terminationTrigger can be set by syscall.exit or a vat-fatal syscall
-    terminationTrigger = null; // reset terminationTrigger before delivery
-
-    // 'deduction' remembers any meter deduction we performed, in case we
-    // unwind state and have to apply it again
-    let deduction;
-    let vatID;
-    let discardFailedDelivery;
-
-    // The DeliveryStatus tells us what happened to the delivery (success or
-    // worker error). It will be null if the delivery got cancelled, like a
-    // 'notify' or 'retireExports' that was superceded somehow.
-
-    const status = await deliverRunQueueEvent(message);
+    // TODO: policyInput=['crank-failed',{}] is meant to cover
+    // deliveries which fail so badly we don't get metering data (but
+    // which might have taken up considerable time
+    // anyways). Previously, this was triggered by meter underflow
+    // (which only happens when we have metering, so the metering data
+    // is more accurate and should be reported directly), or for a
+    // delivery failure (deliveryResult[0]!=='ok') (which means hard
+    // metering fault, bug in liveslots, or startVat() Kind
+    // restoration error). We now get metering for everything but hard
+    // metering faults.
+
+    // The CrankResults tell us what we should do at end-of-crank:
+    // abort/commit the changes, pop the aborted delivery message,
+    // deduct a meter, decrement the vat's reapcount, and maybe
+    // terminate the vat
+
+    const crankResults = await deliverRunQueueEvent(message);
+    // { abort/commit, deduct, terminate+notify, popDelivery }
+
+    // Deliveries cause syscalls, syscalls might cause errors
+    // (Reported through vatFatalSyscall() and the 'illegalSyscall'
+    // flag), those errors require the delivery consequences be
+    // unwound (and the vat either terminated or suspended
+    // somehow). syscall.exit(isfailure=true) requests an
+    // unwind. Failed upgrades also require an unwind and set .abort
+    // directly.
+
+    if (crankResults.abort) {
+      // errors unwind any changes the vat made
+      kernelKeeper.abortCrank();
```

Comment:
I am not sure I manage to wrap my head around the implications of aborting the crank before doing all the other things like decrementing the reapCounter. It looks like meter and popping delivery were explicitly handled after abort, and I suppose reap is never happening if we abort, so it's equivalent?

Reply:
Yeah, previously we had a mix of changes in the crank buffer by the time we decided to `abort` or `commit`: some made by the vat (syscall consequences), some made by the kernel (decrementing reapCounter, incrementing crankNum, deducting meters, pushing vat-admin messages on the run-queue). So if we decided to abort, we had to re-perform any state changes that were supposed to survive (yes crankNum, yes meter deduction, yes vat-admin, but not syscalls, and not reapCounter because the vat is being terminated anyways, or at least the worker state is being unwound so the amount of garbage is not changing).

This new approach makes the `abort` decision early, and then most of the kernel changes are applied *after* the `abortCrank`, so we don't need to remember them in RAM across the abort point. `CrankResults` remembers `.terminate` (and executes it after any possible `abortCrank`) rather than calling `terminateVat` from inside `processSend` (before the abort) and then recording the details for possible re-application after the abort. Meter deduction is performed after the abort.